File: fn/src/main/java/org/sqlbuilder/fn/joins/LexUnit_Governor.java
package org.sqlbuilder.fn.joins;

import org.sqlbuilder.common.Insertable;
import org.sqlbuilder.fn.RequiresIdFrom;
import org.sqlbuilder.fn.objects.Governor;

import java.util.Comparator;
import java.util.HashSet;
import java.util.Set;

public class LexUnit_Governor extends Pair<Integer, Governor> implements Insertable
{
	public static final Comparator<LexUnit_Governor> COMPARATOR = Comparator.comparing(LexUnit_Governor::getFirst).thenComparing(LexUnit_Governor::getSecond, Governor.COMPARATOR);

	public static final Set<LexUnit_Governor> SET = new HashSet<>();

	// C O N S T R U C T O R

	public static LexUnit_Governor make(final int luid, final Governor governor)
	{
		var ug = new LexUnit_Governor(luid, governor);
		SET.add(ug);
		return ug;
	}

	private LexUnit_Governor(final int luid, final Governor governor)
	{
		super(luid, governor);
	}

	// I N S E R T

	@RequiresIdFrom(type = Governor.class)
	@Override
	public String dataRow()
	{
		return String.format("%d,%s", first, second.getSqlId());
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return String.format("[LU-GOV lu=%s governor=%s]", first, second);
	}
}

File: fn/src/main/java/org/sqlbuilder/fn/joins/LexUnit_Governor.kt
package org.sqlbuilder.fn.joins

import org.sqlbuilder.annotations.RequiresIdFrom
import org.sqlbuilder.common.Insertable
import org.sqlbuilder.fn.objects.Governor

class LexUnit_Governor private constructor(
    luid: Int, governor: Governor,
) : Pair<Int, Governor>(luid, governor), Insertable {

    // I N S E R T

    @RequiresIdFrom(type = Governor::class)
    override fun dataRow(): String {
        return "$first,${second.getSqlId()}"
    }

    // T O S T R I N G

    override fun toString(): String {
        return "[LU-GOV lu=$first governor=$second]"
    }

    companion object {

        @JvmField
        val COMPARATOR: Comparator<LexUnit_Governor> = Comparator
            .comparing<LexUnit_Governor, Int> { it.first }
            .thenComparing<Governor>({ it.second }, Governor.COMPARATOR)

        @JvmField
        val SET = HashSet<LexUnit_Governor>()

        @JvmStatic
        fun make(luid: Int, governor: Governor): LexUnit_Governor {
            val ug = LexUnit_Governor(luid, governor)
            SET.add(ug)
            return ug
        }
    }
}
